storage/scheduledmessagerepo.go

```diff
@@ -0,0 +1,328 @@
+	}
+	if err == nil {
+		message.ProducedBy, err = msgRepo.producerRepository.Get(producerID)
+		if nullDispatchedDate.Valid {
+			message.DispatchedDate = nullDispatchedDate.Time
+		}
+	}
+	if loadChannel && err == nil {
+		message.BroadcastedTo, err = msgRepo.channelRepository.Get(channelID)
+	}
+	return message, err
+}
+
+// MarkDispatched marks a scheduled message as dispatched with the current time
+func (msgRepo *ScheduledMessageDBRepository) MarkDispatched(message *data.ScheduledMessage) error {
+	if message.Status != data.ScheduledMsgStatusDispatched {
+		return fmt.Errorf("can only mark scheduled messages with status DISPATCHED; current status - %v", message.Status)
+	}
+
+	err := transactionalSingleRowWriteExec(msgRepo.db, emptyOps, "UPDATE scheduled_message SET status = ?, dispatchedDate = ?, updatedAt = ? WHERE id = ?",
+		args2SliceFnWrapper(message.Status, message.DispatchedDate, time.Now(), message.ID))
+
+	return err
+}
+
+// GetMessagesReadyForDispatch retrieves scheduled messages that are ready to be dispatched
+func (msgRepo *ScheduledMessageDBRepository) GetMessagesReadyForDispatch(limit int) []*data.ScheduledMessage {
+	var messages []*data.ScheduledMessage
+	query := fmt.Sprintf("%s status = ? AND dispatchSchedule <= ? ORDER BY dispatchSchedule ASC LIMIT %d", scheduledMessageSelectRowCommonQuery, limit)
+	log.Debug().Str("query", query).Int("limit", limit).Msg("Querying for scheduled messages ready for dispatch")
+	rows, err := msgRepo.db.Query(query, data.ScheduledMsgStatusScheduled.GetValue(), time.Now())
+	if err != nil {
+		log.Error().Err(err).Msg("Error querying for scheduled messages ready for dispatch")
+		return messages
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var producerID string
+		var channelID string
+		var nullDispatchedDate sql.NullTime
+		message := &data.ScheduledMessage{}
+		err := rows.Scan(&message.ID, &message.MessageID, &producerID, &channelID, &message.Payload, &message.ContentType, &message.Priority, &message.Status, &message.DispatchSchedule, &nullDispatchedDate, &message.Headers, &message.CreatedAt, &message.UpdatedAt)
+		if err != nil {
+			log.Error().Err(err).Msg("Error scanning scheduled message")
+			continue
+		}
+		if nullDispatchedDate.Valid {
+			message.DispatchedDate = nullDispatchedDate.Time
+		}
+		message.ProducedBy, err = msgRepo.producerRepository.Get(producerID)
+		if err != nil {
+			log.Error().Err(err).Str("producerId", producerID).Msg("Error retrieving producer for scheduled message")
+			continue
+		}
+		message.BroadcastedTo, err = msgRepo.channelRepository.Get(channelID)
+		if err != nil {
+			log.Error().Err(err).Str("channelId", channelID).Msg("Error retrieving channel for scheduled message")
+			continue
+		}
+		messages = append(messages, message)
+	}
+	return messages
+}
+
+// GetScheduledMessagesForChannel retrieves scheduled messages for a specific channel with optional status filters
```

Comment:
Should we use `SELECT FOR UPDATE ... WITH SKIP LOCKED` to fetch the rows in `GetScheduledMessagesForChannel`?  The concurrently running schedulers will skip fetching rows that are already fetched by another scheduler.  If we do the `SELECT FOR UPDATE ... WITH SKIP LOCKED` query and creation of messages from scheduled messages in the same transaction, we don't have to acquire locks for each message individually.  Reduced lock contention should result in lower load on the db.

https://dev.mysql.com/blog-archive/mysql-8-0-1-using-skip-locked-and-nowait-to-handle-hot-rows/
https://blog.appsignal.com/2025/05/07/an-introduction-to-solid-queue-for-ruby-on-rails.html#on-performance

Reply:
Discussed offline, will create a separate Github issue for potential implementation (as this is a general change in the how we handle locking, and may be applicable in other parts of the codebase as well)

Reply:
Created #226 Resolving this comment.